src/modules/flow/iio/nodes.c

```diff
@@ -25,39 +25,41 @@
 
 #include <sol-iio.h>
 
-struct gyroscope_data {
+
+#define MAX_CHANNEL         (3)
+
+struct iiodevice_data {
     struct sol_iio_config config;
-    struct sol_direction_vector scale;
-    struct sol_direction_vector offset;
     struct sol_drange_spec out_range;
     struct sol_iio_device *device;
-    struct sol_iio_channel *channel_x;
-    struct sol_iio_channel *channel_y;
-    struct sol_iio_channel *channel_z;
+    uint8_t channel_count;
     bool buffer_enabled : 1;
     bool use_device_default_scale : 1;
     bool use_device_default_offset : 1;
+    struct sol_iio_channel *channels[MAX_CHANNEL];
+    double scales[MAX_CHANNEL];
+    double offsets[MAX_CHANNEL];
 };
 
 static void
 reader_cb(void *data, struct sol_iio_device *device)
 {
     static const char *errmsg = "Could not read channel buffer values";
     struct sol_flow_node *node = data;
-    struct gyroscope_data *mdata = sol_flow_node_get_private_data(node);
+    struct iiodevice_data *mdata = sol_flow_node_get_private_data(node);
     struct sol_direction_vector out = {
         .min = mdata->out_range.min,
         .max = mdata->out_range.max
     };
     bool b;
 
-    b = sol_iio_read_channel_value(mdata->channel_x, &out.x);
+    b = sol_iio_read_channel_value(mdata->channels[0], &out.x);
     if (!b) goto error;
 
-    b = sol_iio_read_channel_value(mdata->channel_y, &out.y);
+    b = sol_iio_read_channel_value(mdata->channels[1], &out.y);
     if (!b) goto error;
 
-    b = sol_iio_read_channel_value(mdata->channel_z, &out.z);
+    b = sol_iio_read_channel_value(mdata->channels[2], &out.z);
     if (!b) goto error;
 
     sol_flow_send_direction_vector_packet(node,
```

Comment:
as @edersondisouza pointed out, this is not right as a supposedly "general" `read_cb` is assuming x/y/z and a direction-vector. The port is also being assumed to be `SOL_FLOW_NODE_TYPE_IIO_GYROSCOPE__OUT_OUT`.

As I wrote in my review, there should be multiple `read_cb`, one per interface, like `read_direction_vector_cb`, `read_color_cb`... These should be set in your extended type (as done with port index) with `"extra_methods": {"read_cb": "read_direction_vector_cb"}` in JSON. Instead of registering `read_cb` as you do now, you register `type->read_cb`.

Each `read_XXX_cb` should use `type->port_out` to use the port index, also defined in `extra_methods`.